### PR TITLE
[GPU] Add support for i16, u16, and u32 element types in remote tensors

### PR DESCRIPTION
**Pull request by jade-cho**
### Details:
 - *Removed host memory data converting for user input/output tensors with data types i16, u16, or u32.*
 - *User tensors can now be directly used as plugin tensors without additional data conversion overhead.*

### Tickets:
 - *156709*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp [hunk 1]
```diff
@@ -326,4 +326,42 @@ void add_required_reorders::run(program& p) {
...
+    for (auto node : p.get_processing_order()) {
+        if (node->is_type<input_layout>()) {
+            if (node->get_output_layout().data_type == data_types::i16 ||
+                node->get_output_layout().data_type == data_types::u16 ||
+                node->get_output_layout().data_type == data_types::u32) {
+                auto supported_data_type = cldnn::element_type_to_data_type(node->get_output_layout().data_type);
+                layout reorder_layout = node->get_output_layout();
+                reorder_layout.data_type = supported_data_type;
+
+                auto new_reorder = std::make_shared<reorder>(node->id() + "_convert_type", node->id(), reorder_layout);
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp:339]: Why is that needed here? I'd expect that `ov::pass::ConvertPrecision` inserts explicit `Convert` operations for inputs since we request conversion from most of integer types to i32.

Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp:339]: The purpose of this PR is not only to support element types i16, u16, and u32, but also to share user tensors allocated as cl_mem directly as plugin tensors. Of course I have tried as you suggested, but when creating primitive for Parameter op and Result op within the GPU Plugin, it converts to the element type supported by the plugin, so adding Convert is meaningless. So, I modified cldnn::input_layout to support those data types and chose to add a reorder at the next. In the case of Result op, it is already a reorder, so no additional processing is needed.

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp:339]: Well, I understand the part which allows these types for input_layout primitive. But my point is that we run ConverPrecision pass already for u16/i16/u32 -> i32 conversion which means that for Parameter/Result ops with these types we'll have explicit Convert operation on the ngraph level which will produce reorder in cldnn graph. So I don't understand why you try to insert yet another reorder in this `add_required_reorders` pass

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp:339]: And the test that you've added doesn't really check that this pass works correctly as original graph contains u16 -> f32 convert right after Parameter op.
image

So I think we need a test with u16 Parameter followed by any op which can process u16 tensor directly (Some elementwise probably), i.e. original graph shouldn't contain any type conversions for a part of the ops.

Comment (jade-cho) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/add_required_reorders.cpp:339]: How about this graph for test?
image
image

File: src/plugins/intel_gpu/src/plugin/ops/result.cpp [hunk 2]
```diff
@@ -30,7 +30,8 @@ static void CreateResultOp(ProgramBuilder& p, const std::shared_ptr<ov::op::v0::
     auto out_format = cldnn::format::get_default_format(out_rank);
 
     auto out_primitive_name = layer_type_name_ID(op);
-    auto out_data_type = cldnn::element_type_to_data_type(convert_to_supported_device_type(op->get_input_element_type(0)));
+    auto out_data_type = convert_to_supported_device_type(op->get_input_element_type(0));
+    out_data_type = out_data_type == ov::element::boolean ? ov::element::u8 : out_data_type;
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/plugin/ops/result.cpp:34]: nit: can it be embedded into `convert_to_supported_device_type`?

**Review by vladimir-paramuzov — APPROVED** [review]
overall, LGTM now
